doc/design/migration_to_python.md

```diff
@@ -0,0 +1,479 @@
+
+16. [sqlfs](https://github.com/sql-machine-learning/sqlflow/tree/400c691470c6503393453d47856913df3365503e/pkg/sqlfs) wraps a database table into a file system interface and is required by the `model` package. We would migrate this package to python together with `model`.
+
+17. [table_writer](https://github.com/sql-machine-learning/sqlflow/tree/400c691470c6503393453d47856913df3365503e/pkg/table_writer) is a utility package for rendering database tables for a UI. We **don't** have to migrate this package to python. It's required by the `step` package and should be move to the `step` directory.
+
+## The Proposed Architecture
+
+### The Proposed Golang Package Structure
+
+Based on the foregoing analysis, the directory structure SQLFlow core code under `sqlflow/pkg` should be:
+
+| Package Name | Modification                                                 |
+| ------------ | ------------------------------------------------------------ |
+| `server`     | Calls `executor` instead of `workflow`.                      |
+| `proto`      | Add a new file: `ir.proto`.                                  |
+| `step`       | Remove `step/feature`. Move `tablewriter` to `step/tablewriter`.|
+| `parser`     | Stay the same.                                               |
+| `ir`         | Remove `ir/feature_column.go`.                               |
+| `executor`   | Rename`sql` to `executor`. Remove `sql/codegen` and `sql/codegen/attribute`. |
+| `database`   | Provide a python wrapper.                                    |
+| `pipe`       | Stay the same.                                               |
+| `log`        | Stay the same.                                               |
+
+### The Proposed Python Package Structure
+
+Add a package named `sqlflow` in [sqlflow/python](https://github.com/sql-machine-learning/sqlflow/tree/400c691470c6503393453d47856913df3365503e/python) to contain the python code imgrated from golang. The directory structure of SQLFlow core code under `sqlflow/python/sqlflow` should be:
+
+| Package/Module Name | Description                                                  |
+| ------------------- | ------------------------------------------------------------ |
+| `platform`          | Modules under this package must implement the same set of functions(`run`, `train`, `evaluate`, `explain`, `predict` and `solve`) with the same signatures (e.g. `train(stmt: Statement)`) |
+| `columns`           | Modules under this package defines functions that can be called by users in a `COLUMN` clause. |
+| `features.py`       | Translate the `COLUMN` clause into calls to functions in the `column` package. |
+| `execute.py`        | Execute a SQLFlow program or a SQLFlow step.                 |
+| `client.py`         | A python client of sqlflow_server.                           |
+| `_database.so`      | Wrapper of the golang package `pkg/database`.                |
```

Comment:
GoMaxcompute use HTTP API, rewrite it using pure Python is not much effort.

Reply:
We can wrap all Go DB driver into one Python module, so, we don't have to rewrite them one by one.

Reply:
1. wrapping Go drivers still need to write Python interface to have `Exec`, `Query` functions and assemble the result in Python.
2. calling HTTP API in Python is simple to implement
3. `_database.so` maybe hard to support different OS since model developers may be able to call the Python API to debug models on their laptop.

Reply:
By the way, we only have to implement Alisa Python API, for maxcompute we have pyodps to use handy.